### ISSUE TITLE
🔥🐛 Excel Online: contextual tab sporadically not appearing (both sideloading and store-based add-in)

### ISSUE THREAD

**Issue by wh1t3cAt1k** [OfficeDev/office-js#3476]
## Your Environment

* Platform [PC desktop, Mac, iOS, Office on the web]: Office on the Web
* Host [Excel, Word, PowerPoint, etc.]: Excel
* Office version number: 16.0.16626.42309
* Operating System: Mac
* Browser (if using Office on the web): Chrome latest
 
## Expected behavior

Contextual tab appears every time you create a new workbook with the side-loaded add-in.

## Current behavior

The ribbon sporadically does not appear.

## Steps to reproduce

Download manifest: [app.velixo.com_addin_manifest.xml.zip](https://github.com/OfficeDev/office-js/files/11940170/app.velixo.com_addin_manifest.xml.zip)

Side-load it into Excel Online. 

Create a new workbook (File -> New). Observe whether "Velixo NX" tab did or did not appear in the ribbon.

Repeat the File -> New process until the tab **does not** appear.

## Context

We cannot reliably run end-to-end tests because of that issue. It may also point to a bigger stability issue inside the framework. 

## VIDEO DEMONSTRATION:

https://www.loom.com/share/851a801b58c147bfba7dec75fc6f614f?sid=9e819811-179e-455b-8a72-71934eb8196a

[Velixo internal tracking item](https://velixo.atlassian.net/browse/RD-211)

**Comment by ZYUN-MSFT** [comment]
Hi @wh1t3cAt1k thanks for reaching us.

We are still investigating this issue with internal teams.\
We will keep tracking this issue and let you know if there are any updates.

**Comment by m-hellesen** [comment]
We will be looking into this shortly! Thanks for reporting.

**Comment by wh1t3cAt1k** [comment]
@m-hellesen I don't know if this is related or can assist with your investigation: I have occasionally seen errors in the console saying that `requestCreateControls` has already been called. 

But I know for a fact that we only call it once during start-up.

Looking forward to an update from you!

**Comment by wh1t3cAt1k** [comment]
@m-hellesen @ZYUN-MSFT I linked another old ticket for your convenience and potentially important context.

Back then, the ribbon did not appear in **store-based version** of the add-in. **That** problem got solved some time ago. 

But intermittently missing ribbon problem remains for **side-loaded** add-ins specifically.

**Comment by m-hellesen** [comment]
Thanks for the new information. I am still looking into this, but since it is sporadic, its a bit harder to trace down. Thanks for your patience.

**Comment by wh1t3cAt1k** [comment]
@m-hellesen please let us know if there is any update on this.

We had to reopen the linked bug: https://github.com/OfficeDev/office-js/issues/3052 - the problem still seems replicable in Excel Online even for store-based add-ins.

The root cause might be the same.

**Comment by gmichaud** [comment]
I'm constantly seeing this problem with Excel Online when using a store based add-in.

**Comment by gmichaud** [comment]
Problem still exists...

**Comment by wh1t3cAt1k** [comment]
I have reformulated the issue title since it affects store-based add-ins as well... this is one of the most irksome issues that our clients have on Excel Online, it sometimes takes multiple reloads of the side panel before the ribbon appears.

[Velixo internal tracking item](https://3.basecamp.com/3496385/buckets/16689259/card_tables/cards/6472698410)

**Comment by wh1t3cAt1k** [comment]
Note that https://github.com/OfficeDev/office-js/issues/3052 might be a duplicate at this stage, I suggest merging it into this item...

Some comments in https://github.com/OfficeDev/office-js/issues/3052#issuecomment-1644122925 contain valuable reproduction steps (by @juliakassina)

**Comment by wh1t3cAt1k** [comment]
<img width="599" alt="image" src="https://github.com/user-attachments/assets/13cce2fb-9588-4ff1-aa91-2eb6405cad35">

As per MSFT's suggestion, we added this delay between `requestCreateControls` and `requestUpdate`, and the problem seems to have disappeared.

Probably `requestCreateControls` returns prematurely, and it needs to wait a little more for the server response!

**Comment by mayurandhare13** [comment]
Hi @wh1t3cAt1k, I want to know, what's the context when you invoke contextual tab? Is it just right after launching the taskpane? Or   is there any specific condition?
https://learn.microsoft.com/en-us/office/dev/add-ins/design/contextual-tabs?tabs=jsonmanifest#specify-the-contexts-when-the-tab-will-be-visible-with-requestupdate
like in this example, when charts show up, contextual tab appears.
Also, is there any condition where you hide the tab as well?

**Comment by wh1t3cAt1k** [comment]
It's right after the add-in startup. There is no condition under which we hide it.

**Comment by mayurandhare13** [comment]
Hi @wh1t3cAt1k , essentially, you don't want to create the tab (default/normal) tab mentioning in the main manifest (the one which launch the addin/taskpane). Instead, you want to show the tab when taskpane is launched?

**Comment by wh1t3cAt1k** [comment]
Yes correct, we have had bad experience with any manifest updates in the past, it always resulted in a surge of support requests due to update prompts or unexpected issues. We require dynamic control over the tab without having to resubmit the manifest...

**Comment by mayurandhare13** [comment]
ohh. got it 👍🏻

**Comment by mayurandhare13** [comment]
@wh1t3cAt1k - One more question.
are you calling `requestCreateControl` after something like Office.onReady then when the taskpane opens then `requestUpdate` call to show the tab?

**Comment by wh1t3cAt1k** [comment]
Request create controls with visible: true in JSON is enough on Desktop platforms. On online, it is not, this is the inconsistency we want to solve.

As a workaround in Online, we await a 2 second delay and then call requestUpdate in Excel Online (you cannot see this workaround because we disabled it for microsoft.com accounts). But this is not reliable either, sometimes due to timing issues ribbon doesn't show anyway.

We want this to work consistently in all platforms, so that we don't have to call requestUpdate in Online.

**Comment by mayurandhare13** [comment]
I understand, there is difference in experience between win32 and Web. I'd like to find out how are you doing it.
So far,
- On Win32 - Tab just shows up because in requestUpdaterData you provided visible property. Same is not true for web version.
- Web - you added 2 seconds delay before calling `requestUpdate` to fix the issue.

Ideally, you want to have same experience with win32 & web.

So, shall I assume that you call the requestCreateControl call in something like "Office.onReady()" ?

**Comment by wh1t3cAt1k** [comment]
Yes we call control creation long after awaiting Office.onReady(), after our own initialization finishes. We definitely await the ready event.

**Comment by mayurandhare13** [comment]
Do you meant to say inside `Office.onReady(async()=>{})` ?

![Image](https://github.com/user-attachments/assets/95cb21ee-e644-4764-bad1-b6e89eb67cd4)

**Comment by mayurandhare13** [comment]
@wh1t3cAt1k , we pushed the fix. Now, experience should be similar to desktop. Meaning you could provide the visible property in requestUpdateData so, you won't have to make another API call `requestUpdate`

**Comment by wh1t3cAt1k** [comment]
Fantastic! Can you let me know the build in which it is available?

**Comment by mayurandhare13** [comment]
Changes are in Prod now.

**Comment by wh1t3cAt1k** [comment]
Tested on build 16.0.18723.42301. The additional `requestUpdate` call is not required anymore, the ribbon seems to appear just from `requestCreateControls`.

Thank you very much!!!
